File: source/jugglinglab/generator/Transitioner.java
// Transitioner.java
//
// Copyright 2020 by Jack Boyce (jboyce@gmail.com)

package jugglinglab.generator;

import java.util.ResourceBundle;
import java.util.StringTokenizer;
import javax.swing.JPanel;

import jugglinglab.util.*;


// Base class for all Transitioner objects, which find transitions between
// two patterns in a given notation.

public abstract class Transitioner {
    static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
    static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

    // The built-in transitioners
    public static final String[] builtinTransitioners = { "Siteswap" };

    public static Transitioner newTransitioner(String name) {
        if (name.equalsIgnoreCase("siteswap"))
            return new SiteswapTransitioner();
        return null;
    }

    public void initTransitioner(String arg) throws JuggleExceptionUser, JuggleExceptionInternal {
        int i, numargs;
        StringTokenizer st = new StringTokenizer(arg, " \n");

        numargs = st.countTokens();
        String args[] = new String[numargs];

        for (i = 0; i < numargs; i++)
            args[i] = st.nextToken();

        initTransitioner(args);
    }

    // return the notation name
    public abstract String getNotationName();

    // return a JPanel to be used by ApplicationPanel in the UI
    public abstract JPanel getTransitionerControl();

    // reset control values to defaults
    public abstract void resetTransitionerControl();

    // use parameters from transitioner control
    public abstract void initTransitioner() throws JuggleExceptionUser, JuggleExceptionInternal;

    // use command line args
    public abstract void initTransitioner(String[] args) throws JuggleExceptionUser, JuggleExceptionInternal;

    // run the transitioner with no limits
    public abstract int runTransitioner(GeneratorTarget t) throws JuggleExceptionUser, JuggleExceptionInternal;

    // run the transitioner with bounds on space and time
    public abstract int runTransitioner(GeneratorTarget t, int num_limit,
                    double secs_limit) throws JuggleExceptionUser, JuggleExceptionInternal;
}

File: src/main/kotlin/jugglinglab/generator/Transitioner.kt
//
// Transitioner.kt
//
// Base class for all Transitioner objects, which find transitions between
// two patterns in a given notation.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.generator

import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser
import javax.swing.JPanel

abstract class Transitioner {
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    fun initTransitioner(arg: String) {
        val args: List<String> = arg.split(' ', '\n').filter { it.isNotEmpty() }
        initTransitioner(args.toTypedArray())
    }

    // return the notation name
    abstract val notationName: String

    // return a JPanel to be used by ApplicationPanel in the UI
    abstract val transitionerControl: JPanel

    // reset control values to defaults
    abstract fun resetTransitionerControl()

    // use parameters from transitioner control
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    abstract fun initTransitioner()

    // use command line args
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    abstract fun initTransitioner(args: Array<String>)

    // run the transitioner with no limits
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    abstract fun runTransitioner(t: GeneratorTarget): Int

    // run the transitioner with bounds on space and time
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    abstract fun runTransitioner(t: GeneratorTarget, numLimit: Int, secsLimit: Double): Int

    companion object {
        // The built-in transitioners
        @Suppress("unused")
        val builtinTransitioners: List<String> = listOf(
            "Siteswap",
            )

        fun newTransitioner(name: String): Transitioner? {
            if (name.equals("siteswap", ignoreCase = true)) {
                return SiteswapTransitioner()
            }
            return null
        }
    }
}
